packages/bull-vault/src/FlashBull.sol

```diff
@@ -1,40 +1,295 @@
+    /// @dev Uniswap oracle address
+    address public immutable oracle;
+
+    address public bullStrategy;
+
+    /// @dev params structs
+    struct FlashWithdrawParams {
+        uint256 bullAmount;
+        uint256 maxEthForSqueeth;
+        uint256 maxEthForUsdc;
+        uint24 wPowerPerpPoolFee;
+        uint24 usdcPoolFee;
+    }
+
+    /// @dev data structs from Uni v3 callback
+    struct FlashDepositCrabData {
+        uint256 ethToDepositInCrab;
+    }
+    struct FlashDepositCollateralData {
+        uint256 crabToDeposit;
+        uint256 ethToLend;
+    }
+    struct FlashWithdrawBullData {
+        uint256 bullToRedeem;
+        uint256 usdcToRepay;
+    }
+    struct FlashSwapWPowerPerpData {
+        uint256 bullToRedeem;
+        uint256 crabToRedeem;
+        uint256 wPowerPerpToRedeem;
+        uint256 usdcToRepay;
+        uint256 maxEthForUsdc;
+        uint256 usdcPoolFee;
+    }
+
+    event FlashWithdraw(uint256 bullAmount);
+
+    constructor(address _bull, address _factory, address _oracle) UniBull(_factory) {
+        bullStrategy = _bull;
+        oracle = _oracle;
+        crab = IBullStrategy(_bull).crab();
+        wPowerPerp = IController(IBullStrategy(bullStrategy).powerTokenController()).wPowerPerp();
+        weth = IController(IBullStrategy(bullStrategy).powerTokenController()).weth();
+        usdc = IController(IBullStrategy(bullStrategy).powerTokenController()).quoteCurrency();
+        ethWSqueethPool = IController(IBullStrategy(bullStrategy).powerTokenController()).wPowerPerpPool();
+        ethUSDCPool = IController(IBullStrategy(bullStrategy).powerTokenController()).ethQuoteCurrencyPool();
+    }
+
+    /**
+     * @notice receive function to allow ETH transfer to this contract
+     */
+    receive() external payable {
+        require(msg.sender == weth || msg.sender == bullStrategy);
+    }
+
+    /**
+     * @notice flash deposit into strategy, providing ETH, selling wSqueeth and dollars, and receiving strategy tokens
+     * @dev this function will execute a flash swap where it receives ETH, deposits, mints, and collateralizes the loan using flash swap proceeds and msg.value, and then repays the flash swap with wSqueeth and USDC
+     * @param _ethToCrab ETH that will be deposited into the crab strategy
+     * @param _minEthFromSqth minimum ETH we will receive from the oSQTH-ETH trade for crab component
+     * @param _minEthFromUsdc minimum ETH we will receive from the USDC-ETH trade for leverage component
+     * @param _poolFee Uniswap pool fee
+     */
+    function flashDeposit(uint256 _ethToCrab, uint256 _minEthFromSqth, uint256 _minEthFromUsdc, uint24 _poolFee) external payable {
+        uint256 share;
+        uint256 crabAmount;
+        uint256 usdcToBorrow;
+        uint256 ethToLend;
+        uint256 wSqueethToMint;
+        {
+            (uint256 ethInCrab, uint256 squeethInCrab) = IBullStrategy(bullStrategy).getCrabVaultDetails();
+            
+            uint256 ethFee;
```

Comment:
I'm not sure we need this _calcWsqueethToMint with fee. If we start with _ethToCrab we should just have

 wSqueethToMint = _ethToCrab*(oSQTH in crab)/(ETH in crab)